### PR TITLE
fix(#1444): improve wake classifier robustness on S21 captures

### PR DESCRIPTION
**Pull request by lokhor**
Refs #1444

Retrains the Stage 3 wake classifier with evidence-backed augmentation and
reference-construction window alignment so the five private #1432 S21
captures (including the previously failing c120-06) all produce activation
candidates at any sub-frame phase, while every committed classifier gate
keeps its accepted result.

## Starting state

- Starting main: `fc04ffe35b0a9936549c58242e5e7f914631ccdc`
- Branch: `fix/1444-classifier-robustness`
- Final branch head: `ebfa2a69c4e43991eaf95eb0a7fa67c7336de310`

## Model hashes (SHA-256)

| Model | Old (committed) | New (this PR) |
|---|---|---|
| hey_jandal.onnx | `11bcdb0d800b3a93449197122bd9fb484c4b8db887364c629f6c975e3e38c206` | `3a920e291662d4b58e10432b5c7f686f00073c45972763d55552200b97f9c4a8` |
| melspectrogram.onnx | `ba2b0e0f8b7b875369a2c89cb13360ff53bac436f2895cced9f479fa65eb176f` (unchanged) | same |
| embedding_model.onnx | `70d164290c1d095d1d4ee149bc5e00543250a7316b59f31d056cff7bd3075c1f` (unchanged) | same |

## Gap identified (Phase 2)

Offline replay of the five private captures with the committed model
reproduced the recorded physical results exactly (max |confidence diff|
<= 1.2e-7; manifest in `scripts/private-acoustic-runs/capture-1432/eval-manifest.json`,
gitignored). The failing capture c120-06 never fired at any classifier-window
offset (max 0.0092) while the four passing trials fired 0.7569-0.9110.

Experiments proved:

- amplitude is NOT the discriminator (c120-05 stays ~0.91 across a 32x gain
  sweep; c120-06 never fires at any gain);
- the classifier fires only inside a razor-thin **sub-frame phase band**:
  shifting c120-06's PCM by 384 samples (24 ms) makes the committed model
  fire at 0.67 - the S21 capture's playback-vs-capture clock alignment lands
  trials at arbitrary phases, and the training pipeline had **zero sub-frame
  phase augmentation** (clips embedded on a fixed 1280-sample chunk grid);
- the committed phase-sweep robustness was 9/13, 10/13, 4/13, 11/13, 1/13
  shifts per capture.

The fix is the smallest training-side change: sub-frame phase jitter
(0-1279 samples) plus peak-normalisation coverage of the measured S21 range
(0.05-0.95 vs the original 0.7-0.95) in the existing positive augmentation,
and reference-OWW window construction (phrase onset at window positions
2-5, soft-positive position 6 pinning the committed [1:17] band, near-phrase
hard negatives at the committed floor alignments).

## Training provenance

- Environment: Python 3.14.6, torch 2.13.0+cpu, torchaudio 2.11.0+cpu, onnx
  1.22.0, onnxruntime 1.28.0, openwakeword 0.4.0 (the version the committed
  fixtures were generated with), numpy 2.5.1, piper-tts 1.6.0; torch seed 42.
- Data: 1428 positives (1248 Piper TTS across 12 medium voices x 13 phonetic
  variants x 5 augmentations; 180 natural_wake fixture-derived real
  recordings) + 990 deterministic synthetic negatives (seed 42, mirroring
  the committed negative fixture classes; int16 truncation matches the
  committed generator bit-for-bit - verified embedding-identical to the
  committed `*_stream.json` fixtures).
- Command:
  `python training/train.py --positive_dir training/wakeword/data/positives_c1 --negative_dir training/wakeword/data/negatives --output_dir training/wakeword/output/c2 --epochs 100 --false_positive_weight 4.5 --batch_size 512 --neg_count 30000 --mix_snr_low 15 --mix_snr_high 25 --soft_positive_label 0.50`
- Full provenance (including per-candidate records): `training/wakeword/output/c2/training-manifest.json`
  (gitignored, on the training machine).
- Architecture unchanged: openWakeWord reference dnn (flatten -> Linear(1536,32)
  -> LayerNorm -> ReLU -> Linear(32,32) -> LayerNorm -> ReLU -> Linear(32,1) ->
  Sigmoid) - the same topology as the committed asset (which is NOT the FCN
  that the old `train.py` docstring described; the script now trains the real
  architecture).
- Candidates: C0 (original recipe on regenerated data - reproduces the gap),
  C1 (reference end-aligned construction - position-invariant, failed
  committed floors), C2 selected.

## Private captures stayed evaluation-only

- The five c120-02..06 PCM streams were never used as training inputs; only
  the phase-sweep/onset analysis and scoring used them.
- No private PCM, transcripts, embeddings or tensor dumps are committed.
- The committed `fixture_stream.json` and negative `*_stream.json` fixtures
  are used only in the JVM tests, not in training data. The natural_wake
  fixture audio (private, gitignored under `training/wakeword/data/`) is used
  as a real-recording positive with augmentation - consistent with the
  original recipe's real-recordings class (the original 29 recordings are not
  recoverable on this machine).

## Baseline vs candidate (private captures, exact production framing)

| Trial | Baseline max conf | Candidate max conf | Baseline >=0.65 | Candidate >=0.65 |
|---|---|---|---|---|
| c120-02 | 0.9042 | 1.0000 | 1 | 18 |
| c120-03 | 0.7673 | 1.0000 | 1 | 49 |
| c120-04 | 0.7569 | 1.0000 | 1 | 15 |
| c120-05 | 0.9110 | 1.0000 | 1 | 26 |
| c120-06 | 0.0092 (FAIL) | 1.0000 | 0 | 36 |

Sub-frame phase sweep (13 shifts, max conf >= 0.65): baseline 9/13, 10/13,
4/13, 11/13, **1/13**; candidate **13/13 for all five trials**.

## Committed fixture results (JVM tests)

- Positives: [2:18] 0.9986, [3:19] 1.0, [4:20] 1.0, [5:21] 0.9992 (all >= 0.65);
  [1:17] 0.3843 in [0.2, 0.5).
- Collapsed alignments: [0:16] < 0.01, [6:22] < 0.01; all trailing windows
  < 0.002.
- Negatives (all windows): silence/white/speech-shaped/formant <= 0.0004,
  pink 0.0019 - all < 0.002 floor.
- New regression: `full phrase onset band reaches the activation band`
  ([2:18]-[5:21] all >= high).
- `WakeWordFeaturePipelineParityTest` reference regenerated with the new
  classifier (`parity_reference.json`, `generate_parity_reference.py` hash
  pin updated).

## Size and latency

- Model size: 205,430 -> 205,404 bytes (-26 B, -0.01%).
- CPU inference latency (JVM onnxruntime 1.24.3, 200 runs): 0.017 -> 0.015 ms
  per [1,16,96] inference - no regression.
- Input [1,16,96], output [1,1], opset 13 - unchanged; no detector code
  change required; continuous inference frequency unchanged.

## Automated validation

- `./gradlew :core:voice:testDebugUnitTest` - 226 tests, 0 failures.
- `./gradlew :app:testDebugUnitTest` - pass.
- `python3 -m unittest discover -s scripts/tests` - 715 tests, OK.
- `./gradlew lint` - pass.
- `git diff --check` - clean.
- CI Build & Test: pass. Docs drift: pass.

## Unchanged

Thresholds (0.65/0.50), verifier routing, NNAPI/CPU provider selection,
detector framing/ring logic, mel/embedding models, tensor shapes, audio
source configuration, fixture/volume/placement/acoustic route - all
unchanged. No PCM preprocessing (no gain/AGC/denoising/EQ) added.

## Bounded physical smoke: FAILED (gate stopped)

One exact APK (head `ebfa2a69`, APK SHA-256
`b6c8d9e04129621c97813db3daa3e34e647ace86dd1a12bb5ed7fded750e2573`,
classifier asset hash `3a920e29...`) installed on both devices, fixtures
re-deployed, monitored preflight approved (volume 9, ~0.6 m, built-in
speaker, BT off, target unplugged). Evidence:
`scripts/private-acoustic-runs/smoke-2026-08-05T16-31-35Z-e622b060/` (private).

| Trial | Direction | Verdict | Chain |
|---|---|---|---|
| smoke-s21target-001 | S23U -> S21 | **PASS** | ACTIVATION_CANDIDATE 0.99990 HIGH -> VERIFIED -> WAKE_CALLBACK -> session -> STT_READY -> CUE_PLAYBACK_STARTED -> STT_ERROR(stt_recognition_failed) -> SESSION_CANCELLED -> DETECTOR_REARMED |
| (between trials) | - | **SPONTANEOUS WAKE** | ACTIVATION_CANDIDATE **0.6712 HIGH** -> VERIFIED_ACTIVATION -> WAKE_CALLBACK -> session i3 -> STT_READY -> CUE_PLAYBACK_STARTED -> STT_ERROR -> SESSION_CANCELLED -> DETECTOR_REARMED g4. No source playback was active (S23U idle; next playback ~128 s later). Occurred 4.2 s after rearm, 2.0 s after STAGE3_READY - the first classification of the freshly re-armed ring. |
| smoke-s21target-002 | S23U -> S21 | **FAIL (acoustic_or_gate_miss)** | Playback chain fired correctly: ACTIVATION_CANDIDATE **0.99957 HIGH** -> VERIFIED -> WAKE_CALLBACK -> STT_READY -> CUE_PLAYBACK_STARTED -> STT_ERROR -> SESSION_CANCELLED -> DETECTOR_REARMED g5 (all present in the device journal, s=52..65). But the trial boundary (captured at s=45, mid spontaneous session i3, generation 3) pinned generation 3, while the playback chain ran in generation 4 - correlation failed, verdict FAILED. Source playback itself completed (`completion_status=completed`). |

Stopped per the zero-valid-failure gate. The S23U-target 3-trial phase was
not run.

### Root-cause analysis (offline, private captures)

The new classifier is over-permissive on non-phrase content:

- Replaying the five private capture PCMs through production framing, the
  candidate scores >=0.65 in 18/49/15/26/36 windows per capture, spread
  across the idle window (clusters from 0 s to 110 s), whereas the committed
  classifier exceeds 0.65 in exactly one window per capture (the playback
  phrase at ~114-126 s) and nowhere else.
- The worst case: the very first window of c120-06 (pure zero-prefill ring,
  RMS = 0.0) scores **0.9973** with the candidate; the committed model scores
  it <0.65. The committed "silence" fixtures pass (<0.002) only because they
  exercise a warm ring (trailing silence), not the cold/re-armed prefill
  state.
- The physical spontaneous wake (0.6712 HIGH, 2.0 s after STAGE3_READY) is
  the same phenomenon on-device: the first classification after rearm, on a
  freshly cleared ring plus ~25 frames of room tone, crossed the 0.65 high
  threshold, was verified, invoked the wake callback, started a voice
  session and played the cue - a product-level false positive during gated
  idle, and the direct cause of trial 2's failed correlation.

The offline gates did not cover this state: they checked reach/margin on the
captures (satisfied trivially by the over-permissive model), the committed
positive/negative fixtures (warm-state synthetic negatives), and phase-sweep
robustness. No ambient-content or re-arm-prefill negative coverage existed.

### Verdict

The classifier as trained is **not acceptable**: it fixes the phase-band
miss (c120-06 fires at every phase) but introduces spontaneous false
positives at the 0.65 high threshold on non-phrase content, including the
re-arm prefill transient. Per the issue gate (zero valid failures), the
physical smoke stops here. No further retraining or re-testing was done in
this session.

Follow-up directions for the next iteration (outside this session):
1. Add cold/re-arm prefill and ambient-room negatives to the training data
   (and to the committed JVM negative fixtures).
2. Consider a stage-2 voiced-frame gate on the first classification after
   rearm (detector change - needs review, outside classifier-only scope).
3. Re-run the same bounded physical smoke.

**No full #1410 release matrix was run.** Do not merge - wait for review.

### CODE REVIEW

**Review by NickMonrad — CHANGES_REQUESTED** [review]
## Review result: changes required

The bounded physical gate correctly stopped. The current classifier candidate must not merge.

### High — the committed candidate introduces product-level false activations

**Location:** `app/src/main/assets/models/wakeword/hey_jandal.onnx` (`3a920e291662d4b58e10432b5c7f686f00073c45972763d55552200b97f9c4a8`)

**Failure scenario:** With no source playback, the S21 produced a verified HIGH activation at `0.6712`, invoked the wake callback, started a voice session and played the cue. Offline replay also shows this model scoring many non-phrase windows above `0.65`, including `0.9973` on the first zero-prefill window of c120-06.

**Why this is blocking:** This violates #1444's false-positive and zero-valid-failure acceptance criteria and would create spontaneous wake sessions during normal idle/re-arm operation.

**Smallest remediation:** Do not merge or preserve this model as the selected asset. Replace it only after a new narrowly scoped classifier candidate passes production-faithful cold/re-arm and ambient-negative gates, followed by the same one-exact-APK bounded smoke with zero valid failures.

### Medium — the offline negative gate does not model the production cold/re-arm state

**Locations:**

- `training/train.py:221-232` (`_reset_af`) and `training/train.py:375-410` (negative-window construction)
- `core/voice/src/test/java/com/kernel/ai/core/voice/WakeWordClassifierModelTest.kt:139-151`
- `training/wakeword/generate_negatives.py`

**Failure scenario:** Training negatives are produced through openWakeWord `AudioFeatures` reset/prefill semantics, and the committed classifier tests use pre-generated warm silence/noise embedding streams. Production resets `WakeWordMelFeatureState` and `WakeWordEmbeddingRingState` and reaches a different cold/re-armed window composition. The result is a green offline suite while the exact production zero-prefill/re-arm state scores near 1.0 and wakes the app.

**Why this is blocking:** The tests and training comments claim alignment coverage that does not cover the state that actually failed, so the regression can recur unnoticed.

**Smallest remediation:** Before retraining, add a privacy-safe negative fixture/test generated through the exact production reset path (`WakeWordMelFeatureState.reset` → Stage 1/2 → `WakeWordEmbeddingRingState`) for zero PCM and deterministic ambient PCM. Assert every complete classifier window remains below the low threshold. Add independent ambient-only recordings as local training negatives; do not train on the held-out #1432 capture trials. Then train one evidence-driven replacement candidate and rerun the existing bounded smoke.

Do not add the proposed Stage-2 first-classification gate in this remediation. The candidate also fires on ambient windows well after re-arm, so a detector gate would mask only one manifestation while leaving the invalid classifier behaviour intact.

CI and Docs Drift are green at this head, but they do not override the failed physical acceptance gate.

**Comment by lokhor** [comment]
Closing unmerged: the C2 candidate was rejected physically (spontaneous HIGH wake at 0.6712 on non-phrase content + trial-correlation failure in the bounded smoke), and C3-C5 plus the W1 width-64 candidate all failed offline acceptance (frozen-ACAV crossings, capture idle/ambient selectivity, phase-sweep and fixture-band regressions). #1444 has moved to a Stage 3 architecture investigation (temporal weight sharing in place of the flattened MLP). This PR therefore contains no mergeable candidate. Branch and worktree stay for the investigation.
